consensus/src/main/kotlin/maru/consensus/validation/BlockValidator.kt

```diff
@@ -72,149 +89,196 @@ class CompositeBlockValidator(
   }
 }
 
-object BlockValidators {
-  val BlockNumberValidator =
-    BlockValidator { block, _, parentBlockHeader ->
-      if (block.beaconBlockHeader.number != parentBlockHeader.number + 1u) {
-        SafeFuture.completedFuture(
-          error(
-            "Block number is not the next block number " +
-              "blockNumber=${block.beaconBlockHeader.number} " +
-              "nextBlockNumber=${parentBlockHeader.number + 1u}",
-          ),
-        )
-      } else {
-        SafeFuture.completedFuture(ok())
+class StateRootValidator(
+  private val stateTransition: StateTransition,
+) : BlockValidator {
+  override fun validateBlock(newBlock: BeaconBlock): SafeFuture<Result<Unit, BlockValidationError>> =
+    stateTransition.processBlock(newBlock).thenApply { stateTransitionResult ->
+      when (stateTransitionResult) {
+        is Ok<BeaconState> -> {
+          val postState = stateTransitionResult.value
+          val stateRootHeader =
+            postState.latestBeaconBlockHeader.copy(
+              stateRoot = BeaconBlockHeader.EMPTY_STATE_ROOT,
+            )
+          val expectedStateRoot = HashUtil.stateRoot(postState.copy(latestBeaconBlockHeader = stateRootHeader))
+          if (!newBlock.beaconBlockHeader.stateRoot.contentEquals(expectedStateRoot)) {
+            error(
+              "State root in header does not match state root " +
+                "stateRoot=${newBlock.beaconBlockHeader.stateRoot.encodeHex()} " +
+                "expectedStateRoot=${expectedStateRoot.encodeHex()}",
+            )
+          } else {
+            ok()
+          }
+        }
+        is Err<StateTransition.StateTransitionError> -> error("State transition failed: ${stateTransitionResult.error}")
       }
     }
+}
 
-  val TimestampValidator =
-    BlockValidator { block, _, parentBlockHeader ->
-      if (block.beaconBlockHeader.timestamp <= parentBlockHeader.timestamp) {
-        SafeFuture.completedFuture(
-          error(
-            "Block timestamp is not greater than previous block timestamp " +
-              "blockTimestamp=${block.beaconBlockHeader.timestamp} " +
-              "parentBlockTimestamp=${parentBlockHeader.timestamp}",
-          ),
-        )
-      } else {
-        SafeFuture.completedFuture(ok())
-      }
+class BlockNumberValidator(
+  private val beaconChain: BeaconChain,
+) : BlockValidator {
+  override fun validateBlock(newBlock: BeaconBlock): SafeFuture<Result<Unit, BlockValidationError>> {
+    val currentState = beaconChain.getLatestBeaconState()
```

Comment:
The use of `beaconChain.getLatestBeaconState` here means this and the other validators can only be used for validation of the next block. Certainly the most common case. I think we are also likely to use it during syncing to validate the block we receive from peers.

The code would be clearer and also more general if we used `beaconChain.getBeaconState(newBlock.beaconBlockHeader.parentRoot)` instead.